### PR TITLE
feat: implement asind intrinsic

### PR DESCRIPTION
**Pull request by parth121101**

### CODE REVIEW

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -3108,6 +3109,62 @@ namespace BesselY0 {
...
+            body.push_back(al, b.Assignment(result, b.Call(s, args, return_type)));
+        }
+
+        ASR::symbol_t *new_symbol = make_ASR_Function_t(fn_name, fn_symtab, dep, args,
+            body, result, ASR::abiType::Source, ASR::deftypeType::Implementation, nullptr);
+        scope->add_symbol(fn_name, new_symbol);
+        return b.Call(new_symbol, new_args, return_type);
+    }
+
+} // namespace BesselY0
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_functions.h:3166]:
```suggestion
} // namespace Asind
```

File: integration_tests/intrinsics_203.f90 [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    if (asind(x) - 7.06527281 > 1e-5) error stop
+
+    print *, asind(y)
+    if (asind(y) - 61.1635399 > 1e-5) error stop
+
+    print *, asind(z)
+    if (asind(z) - 32.8198891 > 1e-5) error stop
+
+    print *, asind(i)
+    if (asind(i) - 18.723412014063609_dp > 1e-5) error stop
```
Comment (certik) [inline integration_tests/intrinsics_203.f90:24]:
```suggestion
    if (asind(i) - 18.723412014063609_dp > 1e-12) error stop
```

File: src/libasr/intrinsic_func_registry_util_gen.py [hunk 3]
```diff
@@ -223,6 +223,12 @@
             "kind_arg": True
         }
     ],
+    "Asind": [
+        {
+            "args": [("real",)],
+            "ret_type_arg_idx": 0
+        }
+    ],
```
Comment (certik) [inline src/libasr/intrinsic_func_registry_util_gen.py:231]: The `asind` should be just `asin`, which we have already implemented. It accepts a double precision argument.

File: src/libasr/pass/intrinsic_functions.h [hunk 4]
```diff
@@ -3108,6 +3109,62 @@ namespace BesselY0 {
...
 } // namespace BesselY0
 
+namespace Asind {
+
+    static ASR::expr_t *eval_Asind(Allocator &/*al*/, const Location &/*loc*/,
+            ASR::ttype_t* /*t1*/, Vec<ASR::expr_t*> &/*args*/, diag::Diagnostics& /*diag*/) {
+        return nullptr;
+    }
+
+    static inline ASR::expr_t* instantiate_Asind(Allocator &al, const Location &loc,
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:3119]: Not needed, just use the existing `asin`.

Comment (certik) [reply src/libasr/pass/intrinsic_functions.h:3119]: My apologies, it is needed.

File: src/libasr/intrinsic_func_registry_util_gen.py (hunk 3 above)
Comment (certik) [reply src/libasr/intrinsic_func_registry_util_gen.py:231]: My bad: I mistook it with `dasin` which should be just `asin`. The function `asind` returns degrees, so we need to implement it like you did.

**Comment by certik** [comment]
Sorry for my confusion above. Yes, this is needed, thanks for implementing these F2023 functions.

File: src/libasr/intrinsic_func_registry_util_gen.py (hunk 3 above)
Comment (parth121101) [reply src/libasr/intrinsic_func_registry_util_gen.py:231]: Yes Ondrej, asind returns the answer in degrees while asin return the inverse of sin in radians.

**Comment by parth121101** [comment]
@Pranavchiku I tried the approach suggested by you. Following is the issue that arises.
```
$ lfortran /home/lfortran/integration_tests/intrinsics_197.f90
Traceback (most recent call last):
  Binary file "/home/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main_impl()
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main()
  File "/home/lfortran/src/bin/lfortran.cpp", line 2389, in ??
    return main_app(argc, argv);
  File "/home/lfortran/src/bin/lfortran.cpp", line 2350, in main_app(int, char**)
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/home/lfortran/src/bin/lfortran.cpp", line 948, in ??
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/home/lfortran/src/lfortran/fortran_evaluator.cpp", line 359, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 9784, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    pass_manager.apply_passes(al, &asr, co.po, diagnostics);
  File "/home/lfortran/src/libasr/pass/pass_manager.h", line 311, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    apply_passes(al, asr, _passes, pass_options, diagnostics);
  File "/home/lfortran/src/libasr/pass/pass_manager.h", line 167, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    _passes_db[passes[i]](al, *asr, pass_options);
  File "/home/lfortran/src/libasr/pass/intrinsic_function.cpp", line 433, in LCompilers::pass_replace_intrinsic_function(Allocator&, LCompilers::ASR::TranslationUnit_t&, LCompilers::PassOptions const&)
    v.visit_TranslationUnit(unit);
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 46990, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    this->visit_symbol(*a.second);
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 5160, in LCompilers::ASR::BaseVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 4871, in ??
    case symbolType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 47001, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_Program(LCompilers::ASR::Program_t const&)
    self().transform_stmts(xx.m_body, xx.n_body);
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 46983, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::transform_stmts(LCompilers::ASR::stmt_t**&, unsigned long&)
    self().visit_stmt(*m_body[i]);
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 5177, in LCompilers::ASR::BaseVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 4903, in ??
    case stmtType::DoLoop: { v.visit_DoLoop((const DoLoop_t &)x); return; }
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 47275, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_DoLoop(LCompilers::ASR::DoLoop_t const&)
    self().transform_stmts(xx.m_body, xx.n_body);
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 46983, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::transform_stmts(LCompilers::ASR::stmt_t**&, unsigned long&)
    self().visit_stmt(*m_body[i]);
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 5177, in LCompilers::ASR::BaseVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_stmt(LCompilers::ASR::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 4909, in ??
    case stmtType::If: { v.visit_If((const If_t &)x); return; }
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 47305, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::ReplaceIntrinsicFunctionsVisitor>::visit_If(LCompilers::ASR::If_t const&)
    self().call_replacer();
  File "/home/lfortran/src/libasr/pass/intrinsic_function.cpp", line 147, in LCompilers::ReplaceIntrinsicFunctionsVisitor::call_replacer()
    replacer.replace_expr(*current_expr);
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 46027, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_RealCompare(down_cast<ASR::RealCompare_t>(x));
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 44610, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_RealCompare(LCompilers::ASR::RealCompare_t*)
    self().replace_expr(x->m_left);
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 45943, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_IntrinsicElementalFunction(down_cast<ASR::IntrinsicElementalFunction_t>(x));
  File "/home/lfortran/src/libasr/pass/intrinsic_function.cpp", line 53, in LCompilers::ReplaceIntrinsicFunctions::replace_IntrinsicElementalFunction(LCompilers::ASR::IntrinsicElementalFunction_t*)
    replace_expr(x->m_args[i]);
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 46031, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_RealBinOp(down_cast<ASR::RealBinOp_t>(x));
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 44627, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_RealBinOp(LCompilers::ASR::RealBinOp_t*)
    self().replace_expr(x->m_left);
  File "/home/lfortran/src/libasr/../libasr/asr.h", line 45943, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceIntrinsicFunctions>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_IntrinsicElementalFunction(down_cast<ASR::IntrinsicElementalFunction_t>(x));
  File "/home/lfortran/src/libasr/pass/intrinsic_function.cpp", line 76, in LCompilers::ReplaceIntrinsicFunctions::replace_IntrinsicElementalFunction(LCompilers::ASR::IntrinsicElementalFunction_t*)
    ASR::expr_t* current_expr_ = instantiate_function(al, x->base.base.loc,
  File "/home/lfortran/src/libasr/../libasr/pass/intrinsic_functions.h", line 3085, in ??
    body.push_back(al, b.Assignment(result, b.Call(s, args, return_type)));
  File "/home/lfortran/src/libasr/asr_builder.h", line 810, in LCompilers::ASRUtils::ASRBuilder::Call(LCompilers::ASR::symbol_t*, LCompilers::Vec<LCompilers::ASR::expr_t*>&, LCompilers::ASR::ttype_t*)
    return ASRUtils::EXPR(ASRUtils::make_FunctionCall_t_util(al, loc,
  File "/home/lfortran/src/libasr/asr_utils.h", line 5170, in ??
    Call_t_body(al, a_name, a_args, n_args, a_dt, nullptr, false, nopass);
  File "/home/lfortran/src/libasr/asr_utils.h", line 5019, in ??
    ASR::ttype_t* orig_arg_type = ASRUtils::type_get_past_allocatable(
  File "/home/lfortran/src/libasr/asr_utils.h", line 157, in ??
    if (ASR::is_a<ASR::Pointer_t>(*f)) {
  File "/home/lfortran/src/libasr/asr.h", line 32, in bool LCompilers::ASR::is_a<LCompilers::ASR::Pointer_t, LCompilers::ASR::ttype_t>(LCompilers::ASR::ttype_t const&)
    return T::class_type == x.type;
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __GI___sigaction()
Segfault: Signal SIGSEGV (segmentation fault) received
```

File: src/libasr/asr_builder.h [hunk 5]
```diff
@@ -1043,6 +1043,23 @@ class ASRBuilder {
             nullptr, nullptr));
     }
 
+    ASR::symbol_t* create_c_func(ASRUtils::ASRBuilder b, std::string c_func_name, SymbolTable* fn_symtab, ASR::ttype_t* return_type, int n_args, Vec<ASR::ttype_t*>& arg_types) {
```
Comment (certik) [inline src/libasr/asr_builder.h:1046]: `b` can be removed, just use `this`.

**Review by certik — APPROVED** [review]

**Pull request merged**